pkg/reconciler/scheduling/singleplacementslice/controller.go

```diff
@@ -0,0 +1,272 @@
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/cluster"
+	schedulingv1alpha1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/scheduling/v1alpha1"
+	schedulingv1alpha1listers "github.com/kcp-dev/kcp/pkg/client/listers/scheduling/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/logging"
+
+	edgeclientset "github.com/kcp-dev/edge-mc/pkg/client/clientset/versioned/cluster"
+	edgev1alpha1informers "github.com/kcp-dev/edge-mc/pkg/client/informers/externalversions/edge/v1alpha1"
+	edgev1alpha1listers "github.com/kcp-dev/edge-mc/pkg/client/listers/edge/v1alpha1"
+	"github.com/kcp-dev/edge-mc/pkg/indexers"
+)
+
+const (
+	ControllerName = "scheduling-singleplacementslice-controller"
+)
+
+func NewController(
+	kcpClusterClient kcpclientset.ClusterInterface,
+	edgeClusterClient edgeclientset.ClusterInterface,
+	singlePlacementSliceInformer edgev1alpha1informers.SinglePlacementSliceClusterInformer,
+	edgePlacementInformer edgev1alpha1informers.EdgePlacementClusterInformer,
+	locationInformer schedulingv1alpha1informers.LocationClusterInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue: queue,
+		enqueueAfter: func(ns *corev1.Namespace, duration time.Duration) {
+			key, err := kcpcache.MetaClusterNamespaceKeyFunc(ns)
+			if err != nil {
+				runtime.HandleError(err)
+				return
+			}
+			queue.AddAfter(key, duration)
+		},
+		kcpClusterClient:  kcpClusterClient,
+		edgeClusterClient: edgeClusterClient,
+
+		singlePlacementSliceLister:  singlePlacementSliceInformer.Lister(),
+		singlePlacementSliceIndexer: singlePlacementSliceInformer.Informer().GetIndexer(),
+
+		locationLister:  locationInformer.Lister(),
+		locationIndexer: locationInformer.Informer().GetIndexer(),
+
+		edgePlacementLister:  edgePlacementInformer.Lister(),
+		edgePlacementIndexer: edgePlacementInformer.Informer().GetIndexer(),
+	}
+
+	indexers.AddIfNotPresentOrDie(locationInformer.Informer().GetIndexer(), cache.Indexers{
+		indexers.ByLogicalClusterPath: indexers.IndexByLogicalClusterPath,
+	})
+
+	indexers.AddIfNotPresentOrDie(singlePlacementSliceInformer.Informer().GetIndexer(), cache.Indexers{
+		indexNameSPSByWorkspace: indexSPSByWorkspace,
+	})
+
+	singlePlacementSliceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.enqueueSinglePlacementSlice,
+		UpdateFunc: func(_, newObj interface{}) { c.enqueueSinglePlacementSlice(newObj) },
+		DeleteFunc: c.enqueueSinglePlacementSlice,
+	})
+
+	locationInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueForLocation,
+			UpdateFunc: func(old, obj interface{}) {
+				oldLoc := old.(*schedulingv1alpha1.Location)
+				newLoc := obj.(*schedulingv1alpha1.Location)
+				if !reflect.DeepEqual(oldLoc.Spec, newLoc.Spec) || !reflect.DeepEqual(oldLoc.Labels, newLoc.Labels) {
```

Comment:
Are you aware of https://github.com/kubernetes/apimachinery/blob/release-1.24/pkg/api/equality/semantic.go#L29 ?  I see some Kubernetes code prefers to use that for comparing API object values.

Reply:
Cool, used this.